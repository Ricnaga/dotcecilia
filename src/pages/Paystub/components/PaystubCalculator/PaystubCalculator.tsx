import {
  CeciliaButton,
  CeciliaCheckbox,
  CeciliaDate,
  CeciliaNumber,
  CeciliaText,
} from '@shared/components';

export type PaystubCalculatorFields = {
  name: string;
  hasUnsanitary: boolean;
  unsanitary: number;
  extraHour: boolean;
  fullExtra: boolean;
  hours: number;
  salary: number;
  initialWorkdayMonth: string;
  lastWorkdayMonth: string;
  vtr: number;
  missingDays: number;
  discountedDays: number;
};

type PaystubCalculatorProps = Record<'values', PaystubCalculatorFields> & {
  onChange: (
    key: keyof PaystubCalculatorFields,
    value: number | string,
  ) => void;
  onChecked: (
    key: keyof Pick<
      PaystubCalculatorFields,
      'hasUnsanitary' | 'extraHour' | 'fullExtra'
    >,
  ) => void;
  onPrint: () => void;
};

export function PaystubCalculator({
  values,
  onChange,
  onChecked,
  onPrint,
}: PaystubCalculatorProps) {
  return (
    <>
      <div className="grid grid-cols-12">
        <p className="text-4xl text-left font-bold mb-4 col-span-6">Cálculos</p>
        <CeciliaButton
          className="col-start-9 col-end-13 mb-4"
          title="Imprimir"
          onClick={onPrint}
        />
      </div>
      <div className="grid grid-cols-12 gap-4">
        <div className="col-span-12">
          <CeciliaText
            placeholder="Digite o nome"
            value={values.name}
            onChange={({ target }) => onChange('name', target.value)}
          />
        </div>
        <div className="col-span-6">
          <CeciliaCheckbox
            label="Insalubridade ?"
            checked={values.hasUnsanitary}
            onChecked={() => onChecked('hasUnsanitary')}
          />
          {values.hasUnsanitary && (
            <div className="mt-3">
              <CeciliaNumber
                currency
                placeholder="Insalubridade"
                value={values.unsanitary}
                onChange={({ target }) =>
                  onChange('unsanitary', target.valueAsNumber)
                }
              />
            </div>
          )}
        </div>
        <div className="col-span-6 grid grid-cols-12 justify-around">
          <div className="col-span-6">
            <CeciliaCheckbox
              label="HE?"
              checked={values.extraHour}
              onChecked={() => onChecked('extraHour')}
            />
          </div>
          {values.extraHour && (
            <>
              <div className="col-span-6">
                <CeciliaCheckbox
                  label="100% ?"
                  checked={values.fullExtra}
                  onChecked={() => onChecked('fullExtra')}
                />
              </div>
              <div className="mt-3 col-span-12">
                <CeciliaNumber
                  placeholder="horas"
                  value={values.hours}
                  onChange={({ target }) =>
                    onChange('hours', target.valueAsNumber)
                  }
                />
              </div>
            </>
          )}
        </div>
        <div className="col-span-6">
          <CeciliaNumber
            currency
            placeholder="salário"
            value={values.salary}
            onChange={({ target }) => onChange('salary', target.valueAsNumber)}
          />
        </div>
        <div className="col-span-6">
          <CeciliaNumber
            placeholder="Dias faltados"
            value={values.missingDays}
            onChange={({ target }) =>
              onChange('missingDays', target.valueAsNumber)
            }
          />
        </div>
        <div className="col-span-6">
          <CeciliaNumber
            currency
            placeholder="Valor diário VTR"
            value={values.vtr}
            onChange={({ target }) => onChange('vtr', target.valueAsNumber)}
          />
        </div>
        <div className="col-span-6">
          <CeciliaNumber
            placeholder="Desconto de dias"
            value={values.discountedDays}
            onChange={({ target }) =>
              onChange('discountedDays', target.valueAsNumber)
            }
          />
        </div>
        <div className="col-span-6">
          <CeciliaDate
            label="Primeiro dia útil"
            value={values.initialWorkdayMonth}
            onChange={({ target }) => {
              onChange(
                'initialWorkdayMonth',
                target.value ? target.value.toString() : new Date().toString(),
              );
            }}
          />
        </div>
        <div className="col-span-6">
          <CeciliaDate
            label="Último dia útil"
            value={values.lastWorkdayMonth}
            onChange={({ target }) =>
              onChange(
                'lastWorkdayMonth',
                target.value ? target.value.toString() : new Date().toString(),
              )
            }
          />
        </div>
      </div>
    </>
  );
}
