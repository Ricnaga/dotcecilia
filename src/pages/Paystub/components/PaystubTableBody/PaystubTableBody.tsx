import {
  CeciliaTableCell,
  CeciliaTableRow,
} from '../../../../shared/components';

export function PaystubTableBody() {
  return (
    <tbody className="border-t border-slate-800">
      <tr>
        <CeciliaTableCell colSpan={3} />
        <CeciliaTableCell colSpan={1} className="text-center">
          Ref.
        </CeciliaTableCell>
        <CeciliaTableCell className="text-center">Vencimento</CeciliaTableCell>
        <CeciliaTableCell colSpan={2} className="text-center">
          Descontos
        </CeciliaTableCell>
      </tr>
      <CeciliaTableRow>
        <CeciliaTableCell colSpan={3}>Salário do mês</CeciliaTableCell>
        <CeciliaTableCell colSpan={1} className="text-center">
          30
        </CeciliaTableCell>
        <CeciliaTableCell className="text-center">-</CeciliaTableCell>
        <CeciliaTableCell colSpan={2} />
      </CeciliaTableRow>
      <CeciliaTableRow>
        <CeciliaTableCell colSpan={3}>Hora extra 60%</CeciliaTableCell>
        <CeciliaTableCell colSpan={1} className="text-center">
          0
        </CeciliaTableCell>
        <CeciliaTableCell className="text-center">-</CeciliaTableCell>
        <CeciliaTableCell colSpan={2} />
      </CeciliaTableRow>
      <CeciliaTableRow>
        <CeciliaTableCell colSpan={3}>Hora extra 100%</CeciliaTableCell>
        <CeciliaTableCell colSpan={1} className="text-center">
          0
        </CeciliaTableCell>
        <CeciliaTableCell className="text-center">-</CeciliaTableCell>
        <CeciliaTableCell colSpan={2} />
      </CeciliaTableRow>
      <CeciliaTableRow>
        <CeciliaTableCell colSpan={3}>Insalubridade</CeciliaTableCell>
        <CeciliaTableCell colSpan={1} className="text-center">
          0
        </CeciliaTableCell>
        <CeciliaTableCell className="text-center">-</CeciliaTableCell>
        <CeciliaTableCell colSpan={2} />
      </CeciliaTableRow>
      <CeciliaTableRow>
        <CeciliaTableCell colSpan={2}>Vtr - (01 à 30/09/22)</CeciliaTableCell>
        <CeciliaTableCell className="text-center ">12,25</CeciliaTableCell>
        <CeciliaTableCell className="text-center ">12,25</CeciliaTableCell>
        <CeciliaTableCell className="text-center ">-</CeciliaTableCell>
      </CeciliaTableRow>
      <CeciliaTableRow>
        <CeciliaTableCell colSpan={3}>Falta dia</CeciliaTableCell>
        <CeciliaTableCell className="text-center">0</CeciliaTableCell>
        <CeciliaTableCell className="text-center">-</CeciliaTableCell>
        <CeciliaTableCell className="text-center" colSpan={2}>
          -
        </CeciliaTableCell>
      </CeciliaTableRow>
      <CeciliaTableRow>
        <CeciliaTableCell colSpan={3}>Adiantamento anterior</CeciliaTableCell>
        <CeciliaTableCell className="text-center" />
        <CeciliaTableCell className="text-center" />
        <CeciliaTableCell className="text-center" colSpan={2}>
          -
        </CeciliaTableCell>
      </CeciliaTableRow>
      <CeciliaTableRow>
        <CeciliaTableCell colSpan={3}>Desconto 6% Vtr</CeciliaTableCell>
        <CeciliaTableCell className="text-center" />
        <CeciliaTableCell className="text-center" />
        <CeciliaTableCell className="text-center" colSpan={2}>
          -
        </CeciliaTableCell>
      </CeciliaTableRow>
      <tr className="border-t border-slate-800">
        <td colSpan={4}>
          <p className="mt-4 border-r border-t border-slate-800">Total:</p>
        </td>
        <td colSpan={1}>
          <p className="mt-4 text-center border-r border-t border-slate-800">
            -
          </p>
        </td>
        <td colSpan={3}>
          <p className="mt-4 text-center border-t border-slate-800">-</p>
        </td>
      </tr>
      <tr className="mb-4 border-y border-slate-800">
        <td colSpan={4}>
          <p className="mb-4 border-b border-r border-slate-800">
            Total líquido:
          </p>
        </td>
        <td colSpan={4}>
          <p className="mb-4 border-b border-slate-800 text-center">-</p>
        </td>
      </tr>
    </tbody>
  );
}
