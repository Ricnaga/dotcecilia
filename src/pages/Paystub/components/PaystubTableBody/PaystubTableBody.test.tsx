import { render } from '@testing-library/react';
import { PaystubCalculatorFields } from '../PaystubCalculator/PaystubCalculator';
import { PaystubTableBody } from './PaystubTableBody';

const mockDate = new Date(2022, 10, 10);

const values: PaystubCalculatorFields = {
  name: '',
  hasUnsanitary: false,
  unsanitary: 0,
  extraHour: false,
  fullExtra: false,
  hours: 0,
  salary: 0,
  initialWorkdayMonth: mockDate,
  lastWorkdayMonth: mockDate,
  vtr: 0,
  missingDays: 0,
  discountedDays: 0,
};

jest.mock('./hooks/usePaystubTableBody', () => ({
  __esModule: true,
  usePaystubTableBody: () => ({
    data: {
      formattedValues: {
        salary: 'mock salary',
        hasUnsanitary: true,
        unsanitary: 'mock unsanitary',
        previousAdvance: 'mock previousAdvance',
        total: 'mock total',
        totalDiscount: 'mock totalDiscount',
        netTotal: 'mock netTotal',
      },
      formattedVtr: {
        vtr: 'mock vtr',
        vtrMonthValue: 'mock vtrMonthValue',
        vtrRef: 1000,
        discountedVtr: 'mock discountedVtr',
        dateRange: 'mock dateRange',
      },
      formattedHours: {
        hours: 10,
        extraHour: true,
        extraHourValue: 'mock extraHourValue',
        fullExtra: true,
        fullExtraValue: 'mock fullExtraValue',
      },
      formattedMissingDays: {
        missingDays: 10,
        totalMissingDays: 'mock totalMissingDays',
      },
    },
  }),
}));

describe('Components: PaystubTableBody', () => {
  it('should render correctly', () => {
    const { container } = render(
      <table>
        <PaystubTableBody values={values} />
      </table>,
    );
    expect(container).toMatchSnapshot();
  });
});
