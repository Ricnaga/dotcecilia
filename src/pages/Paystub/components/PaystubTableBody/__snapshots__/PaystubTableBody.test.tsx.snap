// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: PaystubTableBody should render correctly 1`] = `
<div>
  <table>
    <tbody
      class="border-t border-slate-800"
    >
      <tr>
        <td
          class="border-r border-slate-800 undefined"
          colspan="3"
        />
        <td
          class="border-r border-slate-800 text-center"
          colspan="1"
        >
          Ref.
        </td>
        <td
          class="border-r border-slate-800 text-center"
        >
          Vencimento
        </td>
        <td
          class="border-r border-slate-800 text-center"
          colspan="2"
        >
          Descontos
        </td>
      </tr>
      <tr
        class="border-t border-slate-800 undefined"
      >
        <td
          class="border-r border-slate-800 undefined"
          colspan="3"
        >
          Salário do mês
        </td>
        <td
          class="border-r border-slate-800 text-center"
          colspan="1"
        >
          30
        </td>
        <td
          class="border-r border-slate-800 text-center"
        >
          mock salary
        </td>
        <td
          class="border-r border-slate-800 border-b"
          colspan="2"
        />
      </tr>
      <tr
        class="border-t border-slate-800 undefined"
      >
        <td
          class="border-r border-slate-800 undefined"
          colspan="3"
        >
          Hora extra 60%
        </td>
        <td
          class="border-r border-slate-800 text-center"
          colspan="1"
        >
          10
        </td>
        <td
          class="border-r border-slate-800 text-center"
        >
          mock extraHourValue
        </td>
        <td
          class="border-r border-slate-800 undefined"
          colspan="2"
        />
      </tr>
      <tr
        class="border-t border-slate-800 undefined"
      >
        <td
          class="border-r border-slate-800 undefined"
          colspan="3"
        >
          Hora extra 100%
        </td>
        <td
          class="border-r border-slate-800 text-center"
          colspan="1"
        >
          10
        </td>
        <td
          class="border-r border-slate-800 text-center"
        >
          mock fullExtraValue
        </td>
        <td
          class="border-r border-slate-800 undefined"
          colspan="2"
        />
      </tr>
      <tr
        class="border-t border-slate-800 undefined"
      >
        <td
          class="border-r border-slate-800 undefined"
          colspan="3"
        >
          Insalubridade
        </td>
        <td
          class="border-r border-slate-800 text-center"
          colspan="1"
        >
          0
        </td>
        <td
          class="border-r border-slate-800 text-center"
        >
          mock unsanitary
        </td>
        <td
          class="border-r border-slate-800 undefined"
          colspan="2"
        />
      </tr>
      <tr
        class="border-t border-slate-800 border"
      >
        <td
          class="border-r border-slate-800 undefined"
          colspan="2"
        >
          Vtr - 
          mock dateRange
        </td>
        <td
          class="border-r border-slate-800 text-center"
        >
          mock vtr
        </td>
        <td
          class="border-r border-slate-800 text-center"
        >
          1000
        </td>
        <td
          class="border-r border-slate-800 text-center"
        >
          mock vtrMonthValue
        </td>
        <td
          class="border-r border-slate-800 border-r"
          colspan="2"
        />
      </tr>
      <tr
        class="border-t border-slate-800 undefined"
      >
        <td
          class="border-r border-slate-800 undefined"
          colspan="3"
        >
          Falta dia
        </td>
        <td
          class="border-r border-slate-800 text-center"
        >
          10
        </td>
        <td
          class="border-r border-slate-800 text-center"
        />
        <td
          class="border-r border-slate-800 text-center"
          colspan="2"
        >
          mock totalMissingDays
        </td>
      </tr>
      <tr
        class="border-t border-slate-800 undefined"
      >
        <td
          class="border-r border-slate-800 undefined"
          colspan="3"
        >
          Adiantamento anterior
        </td>
        <td
          class="border-r border-slate-800 text-center"
        />
        <td
          class="border-r border-slate-800 text-center"
        />
        <td
          class="border-r border-slate-800 text-center"
          colspan="2"
        >
          mock previousAdvance
        </td>
      </tr>
      <tr
        class="border-t border-slate-800 undefined"
      >
        <td
          class="border-r border-slate-800 undefined"
          colspan="3"
        >
          Desconto 6% Vtr
        </td>
        <td
          class="border-r border-slate-800 text-center"
        />
        <td
          class="border-r border-slate-800 text-center"
        />
        <td
          class="border-r border-slate-800 text-center"
          colspan="2"
        >
          mock discountedVtr
        </td>
      </tr>
      <tr
        class="border-t border-slate-800"
      >
        <td
          colspan="4"
        >
          <p
            class="mt-4 border-r border-t border-slate-800"
          >
            Total:
          </p>
        </td>
        <td
          colspan="1"
        >
          <p
            class="mt-4 text-center border-r border-t border-slate-800"
          >
            mock total
          </p>
        </td>
        <td
          colspan="3"
        >
          <p
            class="mt-4 text-center border-t border-slate-800"
          >
            mock totalDiscount
          </p>
        </td>
      </tr>
      <tr
        class="mb-4 border-y border-slate-800"
      >
        <td
          colspan="4"
        >
          <p
            class="mb-4 border-b border-r border-slate-800"
          >
            Total líquido:
          </p>
        </td>
        <td
          colspan="4"
        >
          <p
            class="mb-4 border-b border-slate-800 text-center"
          >
            mock netTotal
          </p>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
